File: lego/backend/internal/drivergrant/grant.go
/*
Copyright 2026.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package drivergrant signs one-shot, action-bound calls from the trusted SSH
// gateway to an in-sandbox agent driver. The driver receives only the derived
// Ed25519 public key; the HMAC ticket secret and signing key remain outside the
// sandbox, so same-Pod tenant code cannot launch a credential-bearing turn by
// calling localhost directly.
package drivergrant

import (
	"crypto/ed25519"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"time"

	"github.com/bex-co/bex/lego/backend/internal/hmacticket"
)

const (
	Header     = "X-Bex-Driver-Grant"
	ActionTurn = "turn"
	domain     = "bex-agent-driver-grant/v1"
)

type Claims struct {
	SessionID string `json:"ses"`
	Action    string `json:"act"`
	IssuedAt  int64  `json:"iat"`
	ExpiresAt int64  `json:"exp"`
	Nonce     string `json:"jti"`
}

// PublicKey returns the raw Ed25519 verification key as base64url. Derivation
// is domain-separated from every HMAC ticket use of the shared gateway secret.
func PublicKey(secret []byte) (string, error) {
	if len(secret) == 0 {
		return "", errors.New("driver grant secret is empty")
	}
	public := signingKey(secret).Public().(ed25519.PublicKey)
	return base64.RawURLEncoding.EncodeToString(public), nil
}

// Mint creates a short-lived single-action grant. The driver additionally
// consumes jti once, so retrying or replaying a captured gateway request fails.
func Mint(secret []byte, sessionID string, now time.Time, ttl time.Duration) (string, error) {
	if len(secret) == 0 || sessionID == "" || ttl <= 0 {
		return "", errors.New("invalid driver grant configuration")
	}
	nonce, err := hmacticket.Nonce()
	if err != nil {
		return "", err
	}
	claims := Claims{
		SessionID: sessionID,
		Action:    ActionTurn,
		IssuedAt:  now.Unix(),
		ExpiresAt: now.Add(ttl).Unix(),
		Nonce:     nonce,
	}
	payload, err := json.Marshal(claims)
	if err != nil {
		return "", err
	}
	body := base64.RawURLEncoding.EncodeToString(payload)
	signature := ed25519.Sign(signingKey(secret), []byte(body))
	return body + "." + base64.RawURLEncoding.EncodeToString(signature), nil
}

func signingKey(secret []byte) ed25519.PrivateKey {
	mac := hmac.New(sha256.New, secret)
	_, _ = mac.Write([]byte(domain))
	return ed25519.NewKeyFromSeed(mac.Sum(nil))
}

File: lego/agent-image/driver/src/grant.ts
/*
 * Copyright 2026.
 * Licensed under the Apache License, Version 2.0.
 */

import { createPublicKey, verify } from "node:crypto";
import type { IncomingMessage } from "node:http";

const grantHeader = "x-bex-driver-grant";
const clockSkewSeconds = 5;

interface GrantClaims {
  ses: string;
  act: string;
  iat: number;
  exp: number;
  jti: string;
}

function decodeBase64URL(value: string): Buffer {
  return Buffer.from(value, "base64url");
}

// DriverGrantVerifier consumes gateway-signed turn grants exactly once. The
// sandbox receives only an Ed25519 public key, so tenant code can inspect the
// verifier and call localhost but cannot authorize a credential-bearing turn.
export class DriverGrantVerifier {
  private readonly publicKey;
  private readonly sessionID: string;
  private readonly consumed = new Map<string, number>();

  constructor(publicKey: string, sessionID: string) {
    if (!publicKey || !sessionID) {
      throw new Error("driver turn grants require a public key and session id");
    }
    this.publicKey = createPublicKey({
      key: { kty: "OKP", crv: "Ed25519", x: publicKey },
      format: "jwk",
    });
    this.sessionID = sessionID;
  }

  consume(
    request: IncomingMessage,
    action: "turn" | "snapshot",
    now = Math.floor(Date.now() / 1000),
  ): boolean {
    const token = request.headers[grantHeader];
    if (typeof token !== "string") return false;
    const [body, signature, extra] = token.split(".");
    if (!body || !signature || extra !== undefined) return false;
    if (
      !verify(
        null,
        Buffer.from(body),
        this.publicKey,
        decodeBase64URL(signature),
      )
    ) {
      return false;
    }
    let claims: GrantClaims;
    try {
      claims = JSON.parse(
        decodeBase64URL(body).toString("utf8"),
      ) as GrantClaims;
    } catch {
      return false;
    }
    if (
      claims.ses !== this.sessionID ||
      claims.act !== action ||
      typeof claims.iat !== "number" ||
      typeof claims.exp !== "number" ||
      typeof claims.jti !== "string" ||
      !claims.jti ||
      claims.exp < now - clockSkewSeconds ||
      claims.iat > now + clockSkewSeconds
    ) {
      return false;
    }
    for (const [nonce, expiry] of this.consumed) {
      if (expiry < now - clockSkewSeconds) this.consumed.delete(nonce);
    }
    if (this.consumed.has(claims.jti)) return false;
    this.consumed.set(claims.jti, claims.exp);
    return true;
  }
}
